### PR TITLE
Add dtype argument and check dtype consistency in ExternalSource

### PR DESCRIPTION
**Pull request by banasraf**
Signed-off-by: Rafal <Banas.Rafal97@gmail.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

It adds `dtype` argument in the external source op needed for Triton DALI backend auto-config. It also adds a check for dtype consistency from iteration to iteration.
#### Additional information
- Affected modules and functionalities:

External source - C++ and Python.

- Key points relevant for the review:

Argument handling and warning message phrasing.

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: EXTSRC.15

**JIRA TASK**: DALI-2474

### CODE REVIEW

File: dali/pipeline/operator/builtin/external_source.cc [hunk 1]
```diff
@@ -92,6 +92,11 @@ The memory location must match the specified ``device`` parameter of the operato
 For the CPU, the provided memory can be one contiguous buffer or a list of contiguous Tensors.
 For the GPU, to avoid extra copy, the provided buffer must be contiguous. If you provide a list
 of separate Tensors, there will be an additional copy made internally, consuming both memory
-and bandwidth.)code", false);
+and bandwidth.)code", false)
+  .AddOptionalArg("dtype", R"code(Input data type.
```
Comment (mzient) [inline dali/pipeline/operator/builtin/external_source.cc:96]: I guess it's a topic for further discussion, but I think we should announce that this parameter is going to be mandatory (and possibly start issuing warnings if it isn't provided?).

Comment (banasraf) [reply dali/pipeline/operator/builtin/external_source.cc:96]: I added information to the docstring. 
The warning should go to a separate PR that will also add the argument to all of our internal uses of external source.

**Review by awolant — APPROVED** [review]

**Review by mzient — APPROVED** [review]

**Comment by banasraf** [comment]
!build

**Comment by banasraf** [comment]
!build

**Pull request merged**